Makefile

```diff
@@ -1,94 +1,75 @@
+THRIFTRW_SRC := idl/github.com/uber/cadence/cadence.thrift
+# one or more thriftrw-generated file(s), to create / depend on generated code
+THRIFTRW_OUT := $(THRIFT_GENDIR)/cadence/idl.go
 TEST_ARG ?= -race -v -timeout 5m
 BUILD := ./build
 
-THRIFT_GEN=$(GOPATH)/bin/thrift-gen
-
-define thriftrwrule
-THRIFTRW_GEN_SRC += $(THRIFT_GENDIR)/go/$1/$1.go
-
-$(THRIFT_GENDIR)/go/$1/$1.go:: $2
-	@mkdir -p $(THRIFT_GENDIR)/go
-	$(ECHO_V)thriftrw --plugin=yarpc --pkg-prefix=$(PROJECT_ROOT)/$(THRIFT_GENDIR)/go/ --out=$(THRIFT_GENDIR)/go $2
-endef
-
-$(foreach tsrc,$(THRIFTRW_SRCS),$(eval $(call \
-	thriftrwrule,$(basename $(notdir \
-	$(shell echo $(tsrc) | tr A-Z a-z))),$(tsrc))))
-
-# Automatically gather all srcs
-ALL_SRC := $(shell find . -name "*.go" | grep -v -e Godeps -e vendor \
-	-e ".*/\..*" \
-	-e ".*/_.*")
+$(THRIFTRW_OUT): $(THRIFTRW_SRC) yarpc-install
+	@mkdir -p $(dir $@)
+	@# needs to be able to find the thriftrw-plugin-yarpc bin in PATH
+	@PATH="$(BUILD)" \
+		$(BUILD)/thriftrw \
+		--plugin=yarpc \
+		--pkg-prefix=$(IMPORT_ROOT)/$(THRIFT_GENDIR) \
+		--out=$(THRIFT_GENDIR) \
+		$(THRIFTRW_SRC)
+
+# Automatically gather all srcs.
+# Intentionally ignores .gen folder, depend on $(THRIFTRW_OUT) instead.
+ALL_SRC := $(shell \
+	find . -name "*.go" | \
+	grep -v \
+	-e vendor/ \
+	-e .gen/ \
+	-e build/ \
+)
 
 # Files that needs to run lint, exclude testify mock from lint
 LINT_SRC := $(filter-out ./mock%,$(ALL_SRC))
 
-# all directories with *_test.go files in them
-TEST_DIRS := $(sort $(dir $(filter %_test.go,$(ALL_SRC))))
-
 vendor:
 	glide install
 
-yarpc-install: vendor
-	go get './vendor/go.uber.org/thriftrw'
-	go get './vendor/go.uber.org/yarpc/encoding/thrift/thriftrw-plugin-yarpc'
+yarpc-install: $(BUILD)/thriftrw $(BUILD)/thriftrw-plugin-yarpc
+
+$(BUILD)/thriftrw: vendor/glide.updated
+	go build -o "$@" './vendor/go.uber.org/thriftrw'
+
+$(BUILD)/thriftrw-plugin-yarpc: vendor/glide.updated
+	go build -o "$@" './vendor/go.uber.org/yarpc/encoding/thrift/thriftrw-plugin-yarpc'
 
 clean_thrift:
 	rm -rf .gen
 
-thriftc: clean_thrift yarpc-install $(THRIFTRW_GEN_SRC)
+thriftc: yarpc-install $(THRIFTRW_OUT) copyright
 
 copyright: ./internal/cmd/tools/copyright/licensegen.go
 	go run ./internal/cmd/tools/copyright/licensegen.go --verifyOnly
 
 vendor/glide.updated: glide.lock glide.yaml
 	glide install
-	touch vendor/glide.updated
-
-dummy: vendor/glide.updated $(ALL_SRC)
-	go build -i -o dummy internal/cmd/dummy/dummy.go
-
-test: bins
-	@rm -f test
-	@rm -f test.log
-	@for dir in $(TEST_DIRS); do \
-		go test -race -coverprofile=$@ "$$dir" | tee -a test.log; \
-	done;
+	touch $@
 
-bins: thriftc copyright lint dummy
+$(BUILD)/dummy: vendor/glide.updated $(ALL_SRC)
+	go build -i -o $(BUILD)/dummy internal/cmd/dummy/dummy.go
 
-cover_profile: clean copyright lint vendor/glide.updated
-	@mkdir -p $(BUILD)
-	@echo "mode: atomic" > $(BUILD)/cover.out
+test $(BUILD)/cover.out: thriftc copyright $(BUILD)/dummy $(ALL_SRC)
```

Comment:
`copyright` triggers a re-build since it touches a bunch of files, but if it (and `thriftc`, since it also requires `copyright`) are removed then anything depending on `$(BUILD)/cover.out` will only trigger a re-build when necessary.
meanwhile, because `test` is `.PHONY`,  a `make test` will always trigger a re-run of the tests (minus test caching, but it seems like -coverprofile might be preventing that).